Various changes to cl_khr_defined_builtin_kernel (#1)
pjaaskel/OpenCL-Docs: ext/cl_khr_defined_builtin_kernels.asciidoc

|====
| *Date*     | *Version* | *Description*
| 2022-12-13 | 0.1.0     | First formulation as an extension specification like proposed by Ben Ashbaugh.

> | TODO       | TODO      | TODO

Reviewer: Please describe the updates briefly here.

|====
| *Date*     | *Version* | *Description*
| 2022-12-13 | 0.1.0     | First formulation as an extension specification like proposed by Ben Ashbaugh.

> | 2022-10-23 | 0.2.0     |

Reviewer: You are living in the past :)

This extension requires OpenCL 1.2 or later.

> This extension requires cl_khr_tensor (Note: unpublished draft, work in progress)

Reviewer: Before we get this extension for commenting, perhaps describe in parenthesis briefly what it will bring in.

This extension requires OpenCL 1.2 or later.

This extension requires cl_exp_tensor.

[NOTE]

> cl_exp_tensor is unpublished, work-in-progress

Reviewer: Not really unpublished, it's being edited publicly and will be pushed to the public Khronos repo as a PR.

> 

Reviewer: Please add yourself here.

=== Overview

OpenCL 1.2 specifies a built-in kernel as a kernel that is executed on
an OpenCL device or custom device by fixed-function hardware or in firmware.
Applications can query the built-in kernels supported by a device or custom
device.

Built-in kernels are referred to by a name (a C string) without any
semantics attached to the functionality. The semantics behind the name
is completely device specific, typically documented in vendor-specific
extension specifications.

The goal for this extension is to lower the bar for utilizing hardware
accelerated functions in drivers by providing a library of
well-defined built-in kernel with good coverage for common acceleration needs
and which is designed to easily evolve over time.

The device drivers that implement this extension can freely choose which

> subset of DBKs they implement and advertise to the clients. The

Reviewer: ```suggestion
subset of defined built-in-kernels (DBKs) they implement and advertise to the clients. The
```

the lowest levels of the driver stack can provide significant latency benefits,
especially when combined with the command-buffering mechanism.

However, the built-in kernel abstraction works well only when it is widely adopted by
vendors, and when multiple vendors implement the same definitions. Otherwise
each vendor specifies and implements their own built-in kernels closely matching their
own hardware accelerator properties, resulting in lack of cross-vendor
portability in the API abstraction presented to the upper layers of
heterogeneous computing software stacks.

This extension standardizes a set of well-defined built-in kernels the clients can
call from higher level programming stacks built with different languages
and multiple libraries, possibly mix accelerator calls with calls to software kernel
commands, and rely on the driver stack to optimize the execution (especially

> the synchronization and communication) as a low level heterogeneous task graph.

Reviewer: ..."The heterogeneous task graph can be described using multiple command-queues and optionally cached using the command buffer extension (refer to the extension here)."

It aims to promote the use of built-in kernels as a programming model for hardware accelerated

> functionality, to improve cross-vendor portability of hardware accelerated computing.

Reviewer: ```suggestion
functionality which would improve cross-vendor portability of hardware accelerated computing.
```

=== Add new section X.Y.Z Querying Defined Built-in Kernels

To request a defined built-in kernel to be executed in the given
devices use:

[source,c]
----

> cl_dbk_descriptor clCreateDefinedBuiltInKernelDescriptor(

Reviewer: 'Descriptor' is not needed in the name. Most the CL APIs create 'descriptors'.

```suggestion
cl_kernel clCreateDefinedBuiltInKernel(
```

Reviewer: Also, we want to return cl_kernel, otherwise we cannot enqueue them?

Author: This API is used to query availability of a DBK. A corresponding kernel is created later.

    cl_context context,
    cl_uint num_devices,
    const cl_device_id* device_list,

>     cl_dbk_name kernel_name,

Reviewer: Do we need a special name type here or is C string enough still?

Author: cl_bdk_name is supposed to be an enumeration. Enumeration is cheaper than C strings.

  device_list. _num_devices_ must be non-zero.

* _device_list_ is a pointer to a list of devices that are in
  context. _device_list_ must be a non-NULL value. The defined built-in kernels
  are loaded for devices specified in this list.

* _kernel_name_ is the name of the defined built-in kernel listed in Appendix I.

* _kernel_attributes_ is a pointer to the structure declared in
  description of the kernel in Appendix I. The structure holds
  kernel's attributes.

* _cl_dbk_mode_properties_ is a pointer to a list of defined built-in
  kernel mode properties. The supported mode properties are listed in
  DBK's entry with default settings in Appendix I. It is valid to set
  this argument to NULL in which case default properties apply (if
  any).

*clCreateDefinedBuiltInKernelDescriptor* returns a valid kernel
descriptor on success indicated by _errcode_ret_ which is set to
CL_SUCCESS. Otherwise, the returned object is NULL and the
_errcode_ret_ is set to one of following code:

* CL_DBK_INVALID_ATTRIBUTE if one or more kernel attributes violates
  conditions descried in defined built-in kernel entry in Appendix I.

* CL_DBK_UNAVAILABLE if kernel attributes are valid but the

>   kernel is not supported one of the devices.

Reviewer: ```suggestion
  kernel is not supported on one of the devices.
```

* CL_DBK_UNSUPPORTED_MODE_PROPERTY if _cl_dbk_mode_properties_ includes
  at least one property not listed in DBK's entry.

* CL_DBK_UNMET_MAX_RELATIVE_ERROR if the DBK is available but does not
  meet the requested constraint set by
  CL_DBK_PROPERTY_MAX_RELATIVE_ERROR property.

> * TODO: other error cases.

Reviewer: You can just omit the TODO. We will add more error cases when we come up with them in any case. Unless you know already something is missing?

[cols="2,1,2",stripes=odd]
|===
| *DBK Mode Property* | *Property Value* | *Description*

| CL_DBK_PROPERTY_MAX_RELATIVE_ERROR | float

> a| Request a DBK whose maximum relative error is bounded by the given

Reviewer: Now we have to define the "golden reference" somehow. What is the error relative to?

Author: > What is the error relative to?

It's relative to exact mathematical result --> no golden reference is needed.

value measured in ULPs.

| CL_DBK_PROPERTY_NON_DETERMINISTIC | cl_bool

a| Allow results of the kernel to be non-reproducible. This allows
implementation to switch algorithm of the kernel on each launch for

> possibly better performance.

Reviewer: This is a good idea to include so we can allow autotuning, but perhaps the property is too generic. Somehow we should restrict the autotuning to still stay within the relative error. So perhaps just add that sentence that the algorithm (and thus the perceived error) can change as long as it fits to the defined error bounds.

Author: Not just for autotuning. There are also other factors like alignment of the data and and number of threads that may change the result ([link 1](https://www.intel.com/content/www/us/en/developer/articles/technical/introduction-to-the-conditional-numerical-reproducibility-cnr.html) [link 2](https://sc13.supercomputing.org/sites/default/files/WorkshopsArchive/pdfs/wp140s1.pdf)).

> Somehow we should restrict the autotuning to still stay within the relative error. So perhaps just add that sentence that the algorithm (and thus the perceived error) can change as long as it fits to the defined error bounds.

The error is already bounded by CL_DBK_PROPERTY_MAX_RELATIVE_ERROR.

// Idea from https://pytorch.org/docs/stable/notes/randomness.html#cuda-convolution-benchmarking

| ... | -
a|

> Ideas:

Reviewer: These all are good. Move these to the end. Some extensions have open/resolved questions section, we should add it there.

* accumulation with saturation.
* Finite math only.
* Flush denormals to zero.
* data layout preferences (NHWC for convolution).
|===

=== Add new function to 5.8.1 Creating Program Objects

To create a program with a set of defined built-in kernel use:

[source,c]
----

> cl_program clCreateProgramWithDefinedKernels(

Reviewer: We should retain compatibility of the DBK with software defined kernels. Do we really need this API here since we can request the DBK handles from the devices and just queue them? What is the purpose of this API?

Reviewer: OK. From the latter code sample I get the understanding that the difference to clCreateProgramWithBuiltInKernels() is defining the properties, especially the accuracy relaxations. This is fine.

*clCreateProgramWithDefinedKernels* returns a valid program on success
indicated by _errcode_ret_ which is set to CL_SUCCESS. Otherwise, the
returned object is NULL and the _errcode_ret_ is set to one of
following code:

* TODO.

=== Add new function to 5.9.1 Creating Kernel Objects

To get a kernel handle for a defined built-in kernel in a program use:

[source,c]
----
cl_kernel clCreateDefinedBuiltInKernel(
    cl_program program,
    cl_dbk_descriptor kernel_desc,
    cl_int* errcode_ret);
----

* _program_ is a program object with a successfully built executable.

* _kernel_desc_ is a defined built-in kernel descriptor in the program.

* _errcode_ret_ will return an appropriate error code. If errcode_ret is
  NULL, no error code is returned.

*clCreateDefinedBuiltInKernel* returns a valid non-zero kernel object
 and errcode_ret is set to CL_SUCCESS if the kernel object is created
 successfully. Otherwise, it returns a NULL value with one of the
 following error values returned in _errcode_ret_:

* TODO.

=== Add new appendix "Appendix I - Defined Built-in Kernels" to OpenCL API Specification

This chapter describes standard defined built-in kernels (DBK) with

> well-defined semantics. A conformant devices can report to

Reviewer: ```suggestion
well-defined semantics. Devices can report
```

availability of the built-in kernels listed in this section with

> `clCreateDefinedBuiltInKernelDescriptor` call. The availability of a

Reviewer: It would be cleaner to have a similar device query like there is for any BiK as defined by OpenCL 1.2 and possibly merge the additional DBK features with it.

Reviewer: We should be able to list them without creating handles to them.

Reviewer: Thus, shall we also expand the clGetDeviceInfo() related to BiKs with the minimum accuracy query or just have the baseline query? If latter, we can list the kernels in the default BiK list by name, right?

Author: DBK's availability is determined not just by its name or accuracy limit, but also other mode properties and their mode properties to accommodate vendors with hard HW constraints. Querying a list of DBKs based on few properties doesn't seem useful/helpful.

DBK is determined from the arguments passed to the
`clCreateDefinedBuiltInKernelDescriptor` and unavailability of a DBK
is indicated by CL_DBK_UNAVAILABLE error code.

The general client-side abstraction of the DBKs is similar to a call
to a C function of which implementation is hidden. The device driver

> can invoke one or more physical hardware accelerators combined with

Reviewer: ```suggestion
are free to implement a DBK by invoking one or more coarse and fine grained hardware accelerators combined with
```

firmware to implement the semantics as efficiently as possible.

It is the driver's responsibility to handle efficient synchronization and communication
to the hardware accelerator, the internal accelerator state management and resource sharing
across multiple OpenCL contexts.

> Identical DBKs with identical inputs, are not guaranteed to produce

Reviewer: I think we should start with the positive angle by defining that they are expected to produce results across these aspects within the given error ranges.

identical results:

* across vendors,

* across driver versions and

* across devices.

> 

Reviewer: ...why and in which cases the results might differ and within which limits? It's good to recap it here.

Otherwise, identical results are produced unless:

* otherwise stated in DBK's description or

* the DBK has CL_DBK_PROPERTY_NON_DETERMINISTIC property set to true.

> Two DKBs are considered identical if their descriptors are created

Reviewer: ```suggestion
Two DBK descriptors for a device are considered identical if they are created
```

using identical kernel name, kernel attribute and kernel mode property
arguments.

> ==== Standard Defined Built-in Kernels ====

Reviewer: ```suggestion
==== The Defined Built-in Kernels ====
```

> The following list of recognized defined built-in kernels. It is

Reviewer: ```suggestion
The following is list of recognized defined built-in kernels. It is
```

> expected to grow and update while preserving backwards compatibility.

Reviewer: ```suggestion
expected to be expanded and updated over the versions of this extensions, while preserving backwards compatibility.
```

> Each defined built-in kernel entry is organized as followed:

Reviewer: ```suggestion
Each defined built-in kernel entry is organized as follows:
```

> * *Name*: Name of the defined built-in kernel (an enumeration).

Reviewer: Do we want an enumeration or is a string enough? We could expand the old BiK call and follow its usage style as close as possible?

Author: > Do we want an enumeration or is a string enough?

Enumeration is good enough and cheaper to process than C strings.

> We could expand the old BiK call and follow its usage style as close as possible?

Does not seem viable and could be very cumbersome. If we want to include NN convolution with their many attributes (1D/2D/3D mode, strides, kernel sizes dilation, padding sizes, padding modes, grouping), we would propably end up with lots of DBK kernels with awfully long names (e.g. `conv2d_kernel<x>_<y>_stride<x>_<y>_padding_<x _begin>_<x_end>_<y_begin>_<y_end>_padwith<mode>_group<grouping>...`).

Reviewer: I won't consider "cheapness of processing" as a strong motivator here with today's hardware. Perhaps in the 1990s. The most important aspect is the cleanliness of the API and to try to avoid fragmentation. This will now introduce another API completely separate from the old BiK. How do the old BiKs relate to DBKs in the plan? Be a complete separate feature? We might want to think this through a bit more to avoid confusion.

Author: > This will now introduce another API completely separate from the old BiK. How do the old BiKs relate to DBKs in the plan? Be a complete separate feature?

What are the immediate issues having a separate API?

Reviewer: A new API that is close but not exactly to the old one tends to make the interface less clean and more confusing. If we can avoid that, it's a cleaner API and can merge some functionalities with the basic BiK API, like was the original intention. In the current proposal we need another query to list the DBKs in clGetDeviceInfo(), for example, although the concepts are similar.

For example, a more advanced version of the current version with string identifier, but the configurable aspects abstracted away with the attribute mechanism (clCreateProgramWithBuiltinKernelWithProperties() ?) could work. The base name would be string based and we can add properties that restrict the choice. clGetDeviceInfo() would return the DBK name in case any attribute configuration of the DBK is supported by the device.

Author: I don’t mind changing the enumeration to C string after you pointed out:

> clGetDeviceInfo() would return the DBK name in case any attribute configuration of the DBK is supported by the device.

IIUC, with C string as the type of the name for DBKs, the ‘clGetDeviceInfo()’ may list the available DKBs (in some form) in `CL_DEVICE_BUILT_IN_KERNELS_WITH_VERSION` list. 

> … mechanism (clCreateProgramWithBuiltinKernelWithProperties() ?) could work. 
>
> We can always create multiple programs if we need multiple accuracy levels for some reason.

Does this mean that the `clCreateProgramWithBuiltinKernelWithProperties()` creates a single program with a single BiK or DBK in it?

It probably does, because creating a program with multiple DBK instances by the same name but by different attributes and/or properties would cause multiple definition errors. This could be avoided by assigning an unique (vendor designated) kernel name for each unique DBK and BiK that accepts attributes and properties. A `cl_kernel` handle for a specific DBK/BiK instance can be obtained without knowing the kernel name with the currently proposed `clCreateDefinedBuiltinKernel()`. (but perhaps this should be renamed to something else that also covers BiKs - e.g. `clCreateKernelWithDescriptor()`?).

Reviewer: Could we have a generic `clCreateBuiltinKernelWithAttributes()' which would apply for vendor-specific BiKs (documented in vendor documentation) as well as the DBKs alive? It would create a cl_kernel from a BiK name (which can be also a DBK) and would allow setting the attributes for that cl_kernel instance.

Author: Does the `clCreateBuiltinKernelWithAttributes()` implicitly create a `cl_program` too? Like, could the application query the program the kernel is associated with (`CL_KERNEL_PROGRAM`)?

Reviewer: I think not. I'd compare this is clCreateKernel() which creates a kernel handle which has state (the args). Here the state would also include the tuning parameters/attributes.

Author: > I'd compare this is clCreateKernel() which creates a kernel handle which has state (the args).

That function and other variants takes in `cl_program` handle. So if `clCreateBuiltinKernelWithAttributes()`does not create a program - does it mean the kernel is not built after the call? Do we need separate API to build the kernel and associate the kernel to a program?

Reviewer: I mean that we could use the (already standardized) [clCreateProgramWithBuiltinKernels](https://man.opencl.org/clCreateProgramWithBuiltInKernels.html) API for creating a program with a set of BiKs (DBKs or not) then when we instantiate kernels from that "program" (for BiKs its just a collection of BiK refs?) the cl_kernel has internal state for configuring the accelerator's different aspects. What `clBuildProgram()` does with BiKs I'd leave to the implementation internals, but would not require it to be called again.

a|
[start=1]
. The height.
. The width.
| Name: *khr_matmul_v1*
| *Kernel Attributes*
a|
[start=0]
. int: transpose operation (trans) type for matrix A (0 = none, 1 = transpose, 2 = conjugate transpose)
. int: transpose type for matrix B (0 = none, 1 = transpose, 2 = conjugate transpose)
. float: scalar (alpha) to multiply the matrix multiplication result elements with
. float* (input): matrix A
. int: leading dimension of A (0 = row-major, 1 = column-major)
. float* (input): matrix B
. int: leading dimension of B (0 = row-major, 1 = column-major)
. float: scalar (beta) to multiply the C matrix elements with before adding it to the result
. float* (input&output): matrix C which is added to the matrix multiplication result, and stores the output
. int: leading dimension of C (0 = row-major, 1 = column-major)
4+| OpenCL C Semantics
4+a|
[source,c]
----
__kernel void __khr_blas_gemm_float(
   int transA, int transB, float alpha, const global float *A, int ldA,
   const global float *B, int ldB,
   float beta, global float *C, int ldC) {
   // TBD: An example implementation that can be used for verification
   // and as a fallback SW implementation.
}
----

> 4+| *OpenVX Neural Network Extension Compatible Kernels*

Reviewer: This is a good reminder. We should ensure we can implement OpenVX NN extension nicely with the DBKs to keep compatibility with the other Khronos APIs.

Reviewer: (And also other common compuer vision algorithms.)

> Fields of the `cl_dkb_attributes_matmul_v1` structure:

Reviewer: Let's not have the version in the name unless we already have a v2.

> . cl_tensor_desc_t A: Tensor description for input matrix A.

Reviewer: ```suggestion
. cl_tensor A: Tensor description for input matrix A.
```

Let's follow the OpenCL naming conventions.

Author: TBD after cl_khr_tensor draft.

> * CL_DBK_PROPERTY_MAX_RELATIVE_ERROR: Unset property defaults to positive infinity.

Reviewer: Perhaps we should describe the properties elsewhere as most of them should be generic and usable across multiple DBKs. Then here we could just list them by name.

Author: Properties are already described elsewhere and already are generic and reusable (maybe not clearly). This DBK just names the properties with default values.

> | Name: *khr_leaky_relu_v1*

Reviewer: ```suggestion
| Name: *khr_leaky_relu*
```

. cl_tensor_t out: The output tensor.
| *Description*
a|
Applies operation `alpha * x if x < 0 else x` on all
elements of the `in` tensor.

If target device does not support denormals, then `alpha` is flushed
to zero before the operation is applied.

| *Kernel mode properties*
| N/A
| *Attribute validation rules*
a|
* `shapeof(in) == shapeof(out)`
* `elementof(in) == elementof(out)`
* `alpha` must be a finite value.
|===

==== Launching DBKs from the Device Side ====

DBKs are primarily meant to be launched as kernel commands via
host-side command queues.  Optionally, they can be callable from

> device-side via `enqueue_kernel`:

Reviewer: Also we might expose some of them as "intrinsics" available in OpenCL C similar to the OpenCL-based ASIP flow for TCE from ~2010. Then they'd be either work-item, SG, or WG-wide intrinsics. But this is something for the future. The unsolved question is how to represent such intrinsics in SPIR-V.

TBC. This probably needs device-side function corresponding to
clCreateDefinedBuiltInKernelDescriptor.

==== Sample Code ====

[source,c]
----

> // TBD. Similarly in cl_qcom_ml_ops, tensors have type

Reviewer: Perhaps just kick off the cl_tensor extension as another PR to my repo where we can discuss its contents separately. Also we should have examples of DBKs which do not use cl_tensor but cl_mem or even scalar args.

Author: > Also we should have examples of DBKs which do not use cl_tensor but cl_mem or even scalar args. 

What would be the DBK that wouldn't use `cl_tensor`s and why?

// (cl_channel_type) and a number of dimensions (rank) and dimension
// sizes (shape). Difference over the cl_qcom_ml_ops is that the rank is
// "unlimited".
cl_tensor_desc_t lhs_tensor_desc = TBD;
cl_tensor_desc_t rhs_tensor_desc = TBD;
cl_tensor_desc_t res_tensor_desc = TBD;

cl_dkb_attributes_matmul_v1 matmul_attrs = {
  lhs_tensor_desc, rhs_tensor_desc, res_tensor_desc,
  1, 0 // = Transpose lhs tensor
}

cl_dbk_mode_properties matmul_props = {

>   // Request a matmul implementation that meets this precision.

Reviewer: Thinking of this, we have two use cases: 1) Request a list of DBKs from the device with at least a given precision. 2) Enqueuing a DBK with reduced precision requirements (which can help DBK implementation to go faster). If we combine these two, the handle then includes the reduced precision for all invocations of the DBK with that handle. Should be fine for most use cases, and we can request multiple handles.

  CL_DBK_PROPERTY_MAX_RELATIVE_ERROR, 100, // in ULPs.
}

cl_uint err;
std::vector<cl_dbk_descriptor> kernel_descriptions;
cl_dbk_descriptor matmul_desc =
  clCreateDefinedBuiltInKernelDescriptor(
  context, num_devices, device_list,
  CL_DBK_MATMUL_V1, &matmul_attrs, &matmul_props, &err);

} else if (err == CL_DBK_UNAVAILABLE) {
  // Kernel attributes are valid but the kernel is not supported in at least
  // one of the devices.
} else if (err == CL_DBK_UNMET_MAX_RELATIVE_ERROR) {
  // E.g. Kernel is supported but is not precise enough.
} else if (err == CL_DBK_UNSUPPORTED_MODE_PROPERTY) {
  // cl_dbk_mode_properties has a property not listed in the description of the
  // defined built-in kernel.
} else
  kernel_descriptions.push_back(matmul_desc);

...

> cl_program dbk_lib = clCreateProgramWithDefinedBuiltInKernels(

Reviewer: OK. So the distinction to clCreateProgramWithBuiltInKernels is that we'd now communicate the relaxations in accuracy also here? Sounds fine to me. We can always create multiple programs if we need multiple accuracy levels for some reason.

Author: Not sure what you mean here. It is allowed to have two DBKs which are otherwise identical except for a mode property to live in the same program. IOW, one could build a program with matmul DBKs with different accuracy levels.

  context, kernel_descriptions.size(), kernel_descriptors.data(), err);

...

cl_kernel matmul_kernel = clCreateDefinedBuiltinKernel(
  dkb_lib, matmul_desc, err);

// TBD: allocate space for tensors. Perhaps like cl_qcom_ml_ops: query
// tensor sizes after the final program has been created or after
// command buffer (with DBKs within) is finalized. Implementation
// determines the optimal data layout (opaque to the application) for
// the tensors based on their usage.  Application uses the tensor
// sizes to create cl_mem buffers which are bound to the tensors.

> cl_tensor_t lhs_tensor = TBD;

Reviewer: ```suggestion
cl_tensor lhs_tensor = TBD;
```

Author: TBD after cl_khr_tensor draft.

Author: Actually the suggestion makes sense because it would be consistent with OpenCL's other API object types.

cl_tensor_t rhs_tensor = TBD;
cl_tensor_t res_tensor = TBD;

// Transfer data to input tensors

clSetKernelArg(matmul_kernel, 0, sizeof(cl_tensor_t), &lhs_tensor);
clSetKernelArg(matmul_kernel, 1, sizeof(cl_tensor_t), &rhs_tensor);
clSetKernelArg(matmul_kernel, 2, sizeof(cl_tensor_t), &res_tensor);

clEnqueueNDRangeKernel(cmd_q, matmul_kernel, 0, NULL, NULL, NULL, 0, NULL, NULL);
----

=== Open questions

. Should we enable launching DBKs from the device side without requiring device-side enqueue? The main problem is those with NDRange as they are not simple single-WI helper functions.
+
--
*UNRESOLVED*

--

> . Should the NDRange be used at all in DBKs? It feels sort of unnatural as typically the NDRange is used to imply SPMD parallelism while the hardware/firmware is free to choose whatever parallelism degree to implement the function. On the other hand, similar applies to software kernel launches as the work-items can be executed serially if adhering to barrier semantics.

Reviewer: ```suggestion
. Should the NDRange be used at all in DBKs? It feels sort of unnatural as typically the NDRange is used to imply SPMD parallelism while the hardware/firmware is free to choose whatever parallelization strategy to implement the function. On the other hand, similar applies to software kernel launches as the NDRange-launched work-items can be executed serially if adhering to barrier semantics.
```

+
--
*UNRESOLVED*

--

> . Different accelerators prefer different channel orders (NHWC vs. NCHW...) for the processed data. Should the channel order be passed as a DBK argument (like in the example GEMM's row/column order) or is it better to have different DBK variations for each?

Reviewer: Shouldn't this be abstracted away by the cl_tensor, like we discussed?

Author: TBD later.
